### PR TITLE
Add Mautic 3 upgrade script

### LINKED ISSUE #8819: Write upgrade script for M2 to M3
The latest Mautic 3 beta (https://github.com/mautic/mautic/releases/tag/3.0.0-beta2) doesn't support upgrades from Mautic 2 yet, as there's no migration script in place yet to M3.

This issue is to track work on an upgrade script for M2 to M3.

**Pre-upgrade checks**
- [x] PHP version >= 7.2.21 and <= 7.3999
- [x] Current database driver = `pdo_mysql` or `mysqli` (get from existing Mautic config file)
- [x] MySQL version > 5.7.14 or MariaDB version > 10.1
- [x] Mautic version >= 2.16.3. Mautic 2.16 has PHP 7.3 support, so e.g. if the user is currently using PHP 7.1, they can choose between 7.2 and 7.3 prior to upgrading. Next to that, we are then sure that the user has all database migrations prior to upgrading to 3.0. We'll introduce a 2.16.3 version that adds support for upgrading to M3.
- [x] Check if there is a custom configuration for api_rate_limiter_cache in the local.php file (see [here](https://github.com/mautic/mautic/blob/3.x/UPGRADE-3.0.md#configuration) for details)
- [x] Check if Mautic's root folder is writable by creating + deleting a dummy folder
- [x] Ensure PHP's max_execution_time is at least 240 (4 mins). Try to set it to 240 if it's lower than that, if not raise an error.
- [x] Check if `mysqldump` is available on the system, otherwise trigger a warning that the user should take care of the backup
- [x] Inform user to update cron jobs after upgrade (`app/console` to `bin/console`)
- [x] Check if enough free disk space is available for the upgrade [with PHP's](https://www.php.net/manual/en/function.disk-free-space.php) `disk_free_space()` function. NOTE (tested on server with Plesk): This is not reliable on shared hosting, as it will give the total amount of disk space available, not the disk space that's available just to the user's subscription. Best we can do is to raise a warning, not a hard error
- [x] Check if configured spool folder is empty
- [x] Check if there are custom plugins installed (they need to be disabled/upgraded prior to upgrade). Just to be sure I moved the part where we move back plugins/themes/etc. to the very end of the process
- [x] Check if PHP's `memory_limit` is at least 256M

**Upgrade process**
- [x] Send upgrade telemetry to stats server, upgrade status = `started` (https://github.com/mautic/statsapp/pull/1)
- [x] Do a database backup if `mysqldump` is available
- [x] Run database migrations from 2.x branch, just to make sure we're up to date prior to the 3.x upgrade
- [x] Download Mautic 3 files, unzip to a temporary folder called `mautic-3-temp-files`
- [x] Move all current Mautic 2 files to a temporary folder called `mautic-2-backup-files`
- [x] Update all configuration parameters (e.g. `api_rate_limiter_cache`) based on all parameters mentioned in https://github.com/mautic/mautic/issues/8831
- [x] Move files from `mautic-3-temp-files` to Mautic's root folder
- [x] Move config/themes/images/user dashboards/etc. from `mautic-2-backup-files` back into the "empty" 3.0 folder
- [x] Create cache for 3.x
- [x] Run database migrations
- [x] Generate random hashes for the mautic 2 backup folder and database backup file, so we can leave it there just in case something goes wrong.
- [x] Move .htaccess.m3 to .htaccess after we're done
- [x] Add a note about deleting the backup files (button in UI + CLI switch) to give the user the opportunity to delete backup files easily (after confirming everything works as expected)
- [x] Send upgrade telemetry to stats server, upgrade status = `succeeded` | `failed` (https://github.com/mautic/statsapp/pull/1)

**Misc tasks**
- [x] Add error codes with links to the documentation to allow for better troubleshooting
- [x] Build integration with Mautic's "find updates" command to show the upgrade to M3 available if Mautic version > 2.16.3 and kill switch for upgrades = off (on our server)
- [x] Fix steps for CLI upgrade
- [ ] Enable `mysqldump` DB backup only if user has <10000 contacts in their instance and/or recommend user to perform upgrade via CLI instead
- [x] Double-check other configuration changes in `local.php` between M2 and M3 (it also looks like things like mail driver names have changed, as I'm getting SwiftMailer warnings)
- [x] Write testing steps in this PR (including steps to test custom plugins/themes/etc. migrations)
- [x] Database migrations not working
- [x] Add extensive logging to upgrade process and write it into a log file, so that users can send in their log files
- [x] Pass Symfony console errors to the logfile to help us troubleshoot with the users (currently being written to `NullOutput`, should be `BufferedOutput`)
- [x] Build JSON updates in UI so that users see the progress per step and aren't looking at a blank screen for multiple minutes
- [x] Run database migrations 1 by 1 to prevent PHP timeouts (clean 2.16.2 installation has 136 (!) migrations it has to go through)
- [x] Send upgrade error codes to stats server if applicable, so we can see where users are stuck
- [x] Use mau.tc URL shortener for redirecting to the docs
 
I'll create a new `upgrade_v3.php` which will just be included in M2 installations 2.16.3+, to prevent any potential clashes with the existing `upgrade.php`

### PR DESCRIPTION
**Pull request by dennisameling**
Provides a new upgrade script to allow for upgrading M2 to M3. See https://github.com/mautic/mautic/issues/8819 for details

image

**Steps to test**
Mautic 2.16.3 will add a separate upgrade script for the upgrade to Mautic 3. You will first install the Mautic 2.16.3 beta and then upgrade to Mautic 3.

1. You have three options to test this PR:
**Clone this PR**: clone this PR and go to step 3 (assuming that you have a working 2.x installation already. If you clone in an empty folder, then also follow step 2).
**Clean installation**: Download the custom [2.16.3-beta.zip package](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip)
**Upgrade existing 2.x installation**: Download the custom [2.16.3-beta-update.zip package](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta-update.zip)

2. Install 2.16.3-beta (clean install or run `app/console mautic:update:apply --update-package=2.16.3-beta-update.zip` to upgrade an existing installation)

3. You have two options for testing the upgrade to M3 (please do both if you have time):
**UI upgrade**: go to your Mautic dashboard, you should see a reguler update notification like you're used to. If not, run `app/console mautic:update:find` to force search for new updates.
**CLI upgrade**: run `app/console mautic:update:find` and if the update is found, run `php upgrade_V3.php`
4. You should now be able to start the upgrade. Please share your experience with the upgrade script here!

**Testing with DDEV**
For everyone who's using DDEV, you can take these steps to test:

- Open a terminal
- `mkdir mautic3upgrade`
- `cd mautic3upgrade`
- Run `wget https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip`
- Run `unzip 2.16.3-beta.zip`
- Run `ddev config --webserver-type apache-fpm --php-version 7.3`
- Run `ddev start`, and you should be good to go :rocket:

### CODE REVIEW

**Comment by luizeof** [comment]
Hello @dennisameling, how are you? Is this work still in progress? I think of bundle it on the docker-mautic

**Comment by dennisameling** [comment]
@luizeof yes, hope to have this finished by tomorrow, but partly blocked by https://github.com/mautic/mautic/issues/8831 as the migration of all relevant settings needs to be part of the script as well. Will update this PR/issue as soon as I have more!

**Comment by luizeof** [comment]
could you include this #8834 in your script please?

**Comment by dennisameling** [comment]
I'll need to check if that's something the upgrade script could fix or whether it's a Mautic 3 bug. In the meantime, can you try setting up a fresh M3 instance, then try to set the trusted proxy settings in the UI rather than directly adding it to local.php? It could be that the way it's written in the local.php is slightly different in M3 compared to M2. Feel free to report back the results in #8834 and @ me over there to follow up.

**Comment by luizeof** [comment]
@dennisameling will this completely replace the current upgrade.php and also the updater located on the mautic interface bell?

**Comment by dennisameling** [comment]
@luizeof It's a completely separate (standalone) upgrade script (`upgrade_v3.php`), because of the extensive file system changes in M3. It also has an integration with Mautic's native update functionality. So, the user will get a notification in the UI (notification bell) that the upgrade to M3 is available, and when they click "Start update" they are redirected to `upgrade_v3.php`. When running `app/console mautic:update:find` (CLI) you will be instructed on how to start the `upgrade_v3.php` script. If you want to see what that looks like, see https://github.com/mautic/mautic/issues/8819#issuecomment-640244012 :)

**Comment by dennisameling** [comment]
One thing I just found while testing:

On systems with `open_basedir` active, the following error is shown:

image

```
[2020-06-07 21:31:13] ERR_BUILD_M3_CACHE: Failed to build cache for Mautic 3. All your data has been successfully migrated to Mautic 3, but this error very likely needs to be fixed before you can start using Mautic 3. <br><br>Command output: 
 // Clearing the cache for the prod environment with debug                      
 // true                                                                        


In Kernel.php line 363:
                                                                               
  Warning: file_exists(): open_basedir restriction in effect. File(/var/www/v  
  hosts/composer.json) is not within the allowed path(s): (/var/www/vhosts/mysit 
  e.nl/:/tmp/)                                                  
                                                                               

cache:clear [--no-warmup] [--no-optional-warmers] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

This is probably due to the fact that `composer.json` is not included in the M3 package, but deleted in the upgrade package script. Only the `vendor` folder is left.

The script that is executed in this step is `php bin/console cache:clear --no-interaction --env=prod --no-debug`.

**Review by ohbus — APPROVED** [review]

**Comment by kuzmany** [comment]
Upgrade scripts looks like great of piece software with a lot of hardworking time.

My upgrade test experiences:

I did test on clean installation of 2.16.3

### windows localhost enviroment

#### web

- is_mysqldump_available() return true but backup cannot provide, because windows don't know mysqlbackup
- after comment that part I was able to upgrade successfully  👍 
- just clear cache hit timeout - I increased $minExecutionTime to 900

#### console 

I was not able to do upgrade

### linux enviroment produtions (serverpilot)

#### web

Upgrade successful 👍 


#### console 
I was not able to do upgrade

```
Getting the amount of available Mautic 2 database migrations...
string(59) "Couldn't determine amount of available database migrations."
ERR_MAUTIC_2_MIGRATIONS_FAILED: We couldn't reliably detect the amount of available database migrations.
ERR_MAUTIC_2_MIGRATIONS_FAILED: We couldn't reliably detect the amount of available database migrations.. For more details about this message, see https://docs.mautic.org/en/mautic-3-upgrade/mautic-3-upgrade-errors#err_mautic_2_migrations_failed
```

File: upgrade_v3.php [hunk 1]
```diff
@@ -0,0 +1,1974 @@
...
+            outputJSON('success', 'OK', 'applyV2Migrations', 'Applying Mautic 2 database migrations to ensure all migrations are in place prior to upgrade...');
+
+        case 'applyV2Migrations':
+            writeToLog("Getting the amount of available Mautic 2 database migrations...");
+            // Get the amount of available database migrations
+            list($success, $data) = get_available_migrations();
+
+            if ($success === false) {
+                throwErrorAndWriteToLog(
+                    "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```
Comment (RCheesley) [inline upgrade_v3.php]: I think we should differentiate between being unable to find the migrations and there being a failure with applying them. At the moment we have the same error code for both situations of ERR_MAUTIC_2_MIGRATIONS_FAILED.

```suggestion
                    "ERR_MAUTIC_2_MIGRATIONS_IDENTIFICATION_FAILED",
```

File: upgrade_v3.php [hunk 2]
```diff
@@ -0,0 +1,1974 @@
...
+
+                    if ($availableMigrations > 0) {
+                        writeToLog("Applying Mautic 2 database migrations, " . $availableMigrations . " to go...");
+
+                        // Apply migrations on the 2.x branch just so we're sure that we have all migrations in place.
+                        list($success, $output) = apply_single_migration();
+
+                        if ($success === false) {
+                            throwErrorAndWriteToLog(
+                                "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```
Comment (RCheesley) [inline upgrade_v3.php:458]: I think this error code should be specific that the application of the migration has failed.

```suggestion
                                "ERR_MAUTIC_2_MIGRATIONS_APPLICATION_FAILED",
```

File: upgrade_v3.php (hunk 1 above)
Comment (RCheesley) [reply upgrade_v3.php]: I think they should also apply to the M3 migrations

**Comment by dennisameling** [comment]
Thanks for testing @kuzmany! `mysqldump` can actually work on Windows, as long as it's in your PATH. For me that was `C:\wamp64\bin\mariadb\mariadb10.4.10\bin`. I also had a major bug in my logic, `is_mysqldump_available()` was always returning `true`. Just fixed that in https://github.com/mautic/mautic/pull/8824/commits/c2e0c6d63311588517dfbbe584417853746a2137 😄 will check the other problems tomorrow 👍

**Comment by dennisameling** [comment]
Just fixed the "Getting the amount of available Mautic 2 database migrations..." on CLI upgrades. Also added a check that if the platform is Windows, set the `max_execution_time` to 600 instead of 240. Building cache really is painfully slow on this platform, but with a `max_execution_time` that's high enough, it succeeds without errors. https://github.com/mautic/mautic/pull/8824/commits/ab5dbb2a503ad9731e118a641fcb427abf08273b

Will work on the other problems later today and then release a new `2.16.3-beta.zip` package for testing 👍

File: upgrade_v3.php (hunk 1 above)
Comment (dennisameling) [reply upgrade_v3.php]: Done in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251

File: upgrade_v3.php (hunk 2 above)
Comment (dennisameling) [reply upgrade_v3.php:458]: Not necessary due to the introduction of the `_MIGRATIONS_IDENTIFICATION_FAILED` error codes in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251

**Comment by dennisameling** [comment]
Did a bunch of bug fixes and enhancements yesterday and today (https://github.com/dennisameling/mautic/compare/84983e8...84fcd8a):
- Improve upgrade experience on Windows (higher `max_execution_time`, mysqldump improvements)
- Add warning on Windows that cache creation takes long and users should be patient
- Add memory limit check (needs to be at least 256M, on Windows at least 512M)
- Use mau.tc URL shortener for redirects to the docs
- Add separate error codes `ERR_MAUTIC_2_MIGRATIONS_IDENTIFICATION_FAILED` and `ERR_MAUTIC_3_MIGRATIONS_IDENTIFICATION_FAILED`, these indicate that identifying potential migrations failed
- Display PHP errors on CLI upgrade, to make it easier for folks to troubleshoot
- Move the Mautic 2 to 3 file movements to phase 2 of the CLI upgrade (got errors on Windows prior to fixing that)
- Fix mysqldump for users that don't have a password set on their database
- Add option to remove backup files at the end of the upgrade, which looks like this in the UI:

image

Also updated [2.16.3-beta.zip](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip) and [2.16.3-beta-update.zip](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta-update.zip) on their original URLs with the above fixes, so the script is ready to test again 🎉 

cc @kuzmany @RCheesley

**Comment by RCheesley** [comment]
Awesome, will do some tests tomorrow! Fantastic work on this!

**Review by adiux — APPROVED** [review]
tested with ddev, from clean 2.16.3-beta

**Pull request merged**